vc-trusted-issuer-data-service/vc_til_service.py

```diff
@@ -0,0 +1,305 @@
+            'vp_token': BAE_VC
+        }
+        query_params = {
+            'state': params['state'][0]
+        }
+        auth_response = requests.post(auth_response_endpoint, params=query_params, data=auth_params, allow_redirects=False)
+        try:
+            auth_response.raise_for_status()
+        except HTTPError as e:
+            print('HTTP  ERROR')
+            print(e.request.body)
+            print(e)
+            print(e.response.text)
+            raise PluginError('Error when sending request to authentication_response endpoint')
+
+        # Parse received location header
+        auth_response_url = auth_response.headers.get('location')
+        parsed_url = urlparse(auth_response_url)
+        auth_response_params = parse_qs(parsed_url.query)
+
+        # Get access token JWT
+        token_endpoint = "{}/token".format(verifier_url)
+        token_params = {
+            "grant_type": "authorization_code",
+            "code": auth_response_params['code'][0],
+            "redirect_uri": verifier_url
+        }
+        token_response = requests.post(token_endpoint, data=token_params)
+        try:
+            token_response.raise_for_status()
+        except HTTPError as e:
+            print('HTTP  ERROR')
+            print(e.request.body)
+            print(e)
+            print(e.response.text)
+            raise PluginError('Error when sending request to /token endpoint')
+
+        token_data = token_response.json()
+        if 'access_token' not in token_data:
+            raise PluginError('No access token in response')
+
+        return token_data['access_token']
+        
+    def on_product_suspension(self, asset, contract, order):
+        # TODO: Implement PUT overwrite with expired or empty entry, or using DELETE
+        pass
+
+    def on_product_acquisition(self, asset, contract, order):
+
+        # Build trusted issuer entry payload
+        d_now = datetime.now()
+        valid_from = d_now.strftime('%Y-%m-%dT%H:%M:%SZ')
+        d_to = d_now + timedelta(minutes = asset.meta_info['minutes'])
+        valid_to = d_to.strftime('%Y-%m-%dT%H:%M:%SZ')
+        payload = self._get_create_issuer_payload(asset, order, valid_from, valid_to)
+
+        # Get JWT access token from verifier via AS
+        til_endpoint = asset.meta_info['trusted_issuer_endpoint']
+        token = self._same_device_flow(til_endpoint, payload)
+        
+        # Create trusted issuer at TIL
+        issuer_response = requests.post(til_endpoint, json=payload, allow_redirects=False, headers={
+            'Authorization': 'Bearer ' + token
+        })
+        try:
+            issuer_response.raise_for_status()
+        except HTTPError as e:
+            print('HTTP  ERROR')
+            print(e.request.body)
+            print(e)
+            print(e.response.text)
+            raise PluginError('Error when sending request to /issuer endpoint for creating issuer')
+
+    def get_usage_specs(self):
+        return UNITS
```

Comment:
it looks like variable UNITS is not defined

Reply:
fixed